### ISSUE TITLE
RAG PR 1/8: Fix inverted score/score_type contract + reranker fallback signaling

### ISSUE THREAD

**Issue by zaxbysauce** [zaxbysauce/ragappv3#6]
## Phase 1 of the RAG retrieval-quality fix series (1 of 8)

**Ship first.** This PR is the single largest user-visible correctness win in the series — it is responsible for almost every "Tangential" relevance label users currently see, and for the fact that the reranker's actual output is discarded end-to-end.

---

## Problem

Three tightly coupled bugs conspire to make the relevance label shown to the user **literally inverted** relative to actual match quality:

### Bug 1 — `score` carries cosine distance while `score_type` claims "rerank"

- `backend/app/services/rag_engine.py:610`
  ```python
  score_type = "rerank" if self.reranking_enabled else "distance"
  ```
  `score_type` is derived from the config flag, not from whether reranking actually ran.

- `backend/app/services/rag_engine.py:479`
  ```python
  RAGSource(..., score=r.get("_distance", 0.0), ...)
  ```
  The `score` field is always the cosine `_distance` (lower-is-better), never the `_rerank_score` that was just computed.

- `backend/app/services/document_retrieval.py:418`
  ```python
  "score": chunk.score,
  ```
  Passed through to the UI unchanged.

### Bug 2 — `_rerank_score` is computed and discarded

- `backend/app/services/reranking.py:96-101` attaches `_rerank_score` on success. Nothing downstream reads it.

### Bug 3 — Reranker silent fallback looks like success

- `backend/app/services/reranking.py:91-93`
  ```python
  except Exception as e:
      logger.error(f"Reranking failed, returning original order: {e}")
      return chunks[:n]
  ```
  On failure: returns raw dense-ordered chunks without `_rerank_score`, with no success flag. Caller cannot distinguish success from fallback.

### Bug 4 — BGE reranker logits are not sigmoid-normalized

- `backend/app/services/reranking.py:132, 153` return raw logits (e.g., `-2.5` to `+8.0`). Frontend thresholds assume `[0,1]`.

### Bug 5 — Frontend inverts polarity

- `frontend/src/lib/relevance.ts:22-27`
  ```ts
  if (scoreType === "rerank") {
    if (score >= 0.7) return "Highly Relevant";
    if (score >= 0.4) return "Relevant";
    if (score >= 0.2) return "Related";
    return "Tangential";
  }
  ```
  Higher-is-better thresholds. But backend sends cosine distance (lower-is-better).

### Observable effect

| True match quality | Cosine distance sent as `score` | UI label (interpreted as "rerank") |
|---|---|---|
| Near-perfect match | ~0.10 | **Tangential** (score < 0.2) |
| Good match | ~0.30 | Related |
| Mediocre | ~0.50 | Relevant |
| Almost unrelated | ~0.80 | **Highly Relevant** |

Every near-perfect result the user sees is labeled "Tangential." This is the primary driver of the reported symptom "results are almost always listed as tangential rather than related."

### Distance-threshold coupling (Bug 6)

- `backend/app/services/document_retrieval.py:127-151`
  ```python
  skip_distance_filter = reranked or any("_rerank_score" in r for r in results)
  ```
  Bypasses the `MAX_DISTANCE_THRESHOLD` filter when any chunk has `_rerank_score`. Correct in theory but fragile — it couples threshold gating to a flag that Bug 3 already makes unreliable.

---

## Proposed solution

Treat this as one cohesive "scoring contract" fix. Change the contract from implicit to explicit.

### 1. `reranking.py`

- Return `(chunks: List[dict], success: bool)` from `RerankingService.rerank`, OR raise a typed `RerankerUnavailable` exception on failure (pick one; prefer the explicit tuple).
- Always attach a **sigmoid-normalized** `_rerank_score` in `[0,1]` on success: `1.0 / (1.0 + math.exp(-logit))`. Normalize in both `_rerank_via_endpoint` and `_rerank_local`.
- Fallback path attaches nothing (no `_rerank_score`) and signals `success=False`.

### 2. `rag_engine.py`

- Call-site (around L459-471) consumes the success flag:
  - On success: `score_type = "rerank"`, `score = _rerank_score` (already in `[0,1]`).
  - On fallback or reranking disabled: `score_type = "distance"`, `score = _distance`.
- Remove the hardcoded `score_type = "rerank" if self.reranking_enabled else "distance"` line at `rag_engine.py:610`; derive from actual success.
- Fix the token-packing block at `rag_engine.py:473-510` to construct `RAGSource` using the correct field per `score_type`.

### 3. `document_retrieval.py`

- Change `filter_relevant` signature to take an explicit `reranked: bool` instead of sniffing `_rerank_score`. Remove the `any("_rerank_score" in r ...)` heuristic.
- Document `to_source_metadata` contract in a docstring.

### 4. `frontend/src/lib/relevance.ts`

- Confirm the `rerank` thresholds match a sigmoid-normalized BGE distribution. Empirically BGE-reranker-v2-m3 after sigmoid tends to produce clear separation; current thresholds (0.7 / 0.4 / 0.2) are reasonable but validate against real data.
- No polarity change — once the backend sends the correct value, the existing `distance` branch and `rerank` branch of `getRelevanceLabel` both become correct.

### 5. Observability (folded in here because the consumers of this telemetry are downstream PRs)

Add to `retrieval_debug` in the done message (`rag_engine.py:_build_done_message`):
- `rerank_status: "ok" | "fallback" | "disabled"`
- `score_type: "rerank" | "distance"`

---

## Files touched

- `backend/app/services/reranking.py`
- `backend/app/services/rag_engine.py`
- `backend/app/services/document_retrieval.py`
- `frontend/src/lib/relevance.ts` (+ `relevance.test.ts`)

Expected diff: ~200-400 lines.

---

## Tests

Unit:
- Reranker success: every returned chunk has `_rerank_score` in `[0,1]`.
- Reranker failure: returns `(chunks[:n], success=False)`, no `_rerank_score`, original dense order.
- `rag_engine.query()` with reranker success → response has `score_type="rerank"`, `score ∈ [0,1]`.
- `rag_engine.query()` with reranker fallback → response has `score_type="distance"`, `score = _distance`.
- `filter_relevant(..., reranked=True)` skips the distance filter; `reranked=False` applies it.

Integration:
- End-to-end: feed a known near-perfect match, assert `getRelevanceLabel` renders "Highly Relevant" (not "Tangential").
- End-to-end: feed a known near-unrelated result, assert it renders "Tangential" (not "Highly Relevant").

Frontend:
- `relevance.test.ts` parametrized cases for both score types.

---

## Risk & rollout

- **Risk: low.** Types only, no retrieval-ranking change.
- **Visible effect: large.** Users will see a dramatic shift in relevance labels (mostly upward for good matches). This is correct behavior — communicate it in the release notes.
- No feature flag needed; this is a correctness fix.
- No migration, no schema change.

---

## Dependencies

- **Blocks:** PR 5 (fusion tuning), PR 6 (prompt packing), PR 7 (parent-doc retrieval). All later ranking changes rely on this telemetry to be measurable.
- **Independent of:** PR 2, PR 3, PR 4.

---

## Acceptance criteria

- [ ] `_rerank_score` is sigmoid-normalized and sent to the UI as `score` when reranking succeeded.
- [ ] `score_type` reflects the *actual* retrieval path, not the config flag.
- [ ] Reranker fallback is explicitly signaled and does not bypass the distance threshold.
- [ ] `retrieval_debug.rerank_status` is populated on every response.
- [ ] A perfect-match smoke test renders "Highly Relevant" in the UI.
- [ ] All existing unit/integration tests pass.

**Comment by zaxbysauce** [comment]
## Resolution

PR #15 has been merged, addressing all items in this issue.

### Completed Items
- **Reranking contract fix**: `reranking.py` now returns `(chunks, bool)` tuple consistently
- **Sigmoid normalization**: Added `_normalize_logits_to_probabilities()` with overflow guards
- **Rerank fallback signaling**: Explicit `rerank_success` boolean propagated through `_execute_retrieval`
- **8-tuple return**: `rag_engine.py` now returns full debug tuple including `rerank_success`, `score_type`, `hybrid_status`, `fts_exceptions`, `rerank_status`

### Review Fixes Applied
- Log injection vulnerabilities fixed in `reranking.py` and `vector_store.py`
- `_build_done_message` test calls corrected (7 tests fixed)
- Single-chunk bypass correctly returns `success=False`

Closing this issue as completed.
